{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Speeding up the god forsakenly slow d_G_exchange function.\n",
    "\n",
    "### Testing Environment\n",
    "\n",
    "To Standarize the testing of all of these methods, the inputs given to the d_G_exchange function by the first guide generated by the current tool using `GFP.fasta` as an input and the `E_coli_Genome.fasta` and `GFP.fasta` as the output. This should serve as a good use case for meta genomic processing. I'm also making an assumption that if new tools are faster it is worth importing them. \n",
    "\n",
    "### Methods to Test out and implement:\n",
    "\n",
    "* Parallelization \n",
    "* Iteration \n",
    "* Numpy/Pandas to perform tabular functions\n",
    "* Map \n",
    "\n",
    "These are some optimizations we can make towards the end:\n",
    "\n",
    "* Numba\n",
    "* Pypy\n",
    "* Cython\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import numpy as np\n",
    "import scipy.io\n",
    "from itertools import repeat\n",
    "import numba as nb\n",
    "from numba.decorators import jit, autojit, njit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__CODE SETUP__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Variable Files: \n",
    "f = open('Cas_Var.pckl', 'rb')\n",
    "genome_dictionary = pickle.load(f)\n",
    "f.close()\n",
    "\n",
    "mat = scipy.io.loadmat('InvitroModel.mat')\n",
    "weights = mat['w1']\n",
    "\n",
    "#Guide\n",
    "guide = 'CAAACTCAAGAAGGACCATG'\n",
    "\n",
    "#Functions\n",
    "def QuickCalc_Exchange_Energy(crRNA,TargetSeq):\n",
    "    nt_pos={'A':0,'T':1,'C':2,'G':3,'a':0,'t':1,'c':2,'g':3}\n",
    "    dG=0\n",
    "    RNA=''\n",
    "    DNA=''\n",
    "    nt_mismatch_in_first8=0\n",
    "    for i in range(0,len(crRNA)):\n",
    "        pos=20-i\n",
    "        w1=weights[pos]\n",
    "        if nt_pos[crRNA[i]]==nt_pos[TargetSeq[i]]:\n",
    "            dG1=0\n",
    "        else:\n",
    "            # using a bioinformatics search approach to find sequences with up to x mismatches\n",
    "            dG1=2.3 # kcal/mol\n",
    "            if pos<=8:\n",
    "                nt_mismatch_in_first8=nt_mismatch_in_first8+1\n",
    "        dG+=w1*dG1\n",
    "    return float(dG)\n",
    "\n",
    "PAM_energy={'GGA':-9.8,'GGT':-10,'GGC':-10,'GGG':-9.9,'CGG':-8.1,'TGG':-7.8,'AGG':-8.1,'AGC':-8.1,'AGT':-8.1,'AGA':-7.9,'GCT':-7.1,'GCG':-6.9,'ATT':-7.2,'ATC':-6.4,'TTT':-7.6,'TTG':-6.8,'GTA':-7.4,'GTT':-7.9,'GTG':-7.7,'AAT':-7,'AAG':-7,'TAT':-7.2,'TAG':-7.2,'GAA':-7.2,'GAT':-7.3,'GAC':-7.2,'GAG':-7.3}\n",
    "\n",
    "def returnAllPAMs():\n",
    "\n",
    "    for (PAMpart,energies) in sorted(list(PAM_energy.items()), key = lambda x: x[1]):  #PAMpart will be 'GGT'\n",
    "        for nt in ('A','G','C','T'):        #nt + PAMpart will be all possible 'NGGT'\n",
    "            yield nt + PAMpart\n",
    "            \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Current Code__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 430,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2min 41s ± 15.7 s per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit \n",
    "for (source, targets) in genome_dictionary.items():\n",
    "    for fullPAM in returnAllPAMs():\n",
    "        dG_PAM = 0\n",
    "        dG_supercoiling= 0\n",
    "        for (target_sequence, targetPosition) in genome_dictionary[source][fullPAM]:\n",
    "            dG_exchange = QuickCalc_Exchange_Energy(guide, target_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Current Loop__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 431,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "130 ms ± 2.27 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit \n",
    "for (source, targets) in genome_dictionary.items():\n",
    "    for fullPAM in returnAllPAMs():\n",
    "        dG_PAM = 0\n",
    "        dG_supercoiling= 0\n",
    "        for (target_sequence, targetPosition) in genome_dictionary[source][fullPAM]:\n",
    "            dG_exchange = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__New Loop__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 432,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.31 s ± 128 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit \n",
    "for (source, targets) in genome_dictionary.items():\n",
    "    dG_dictionary = {}\n",
    "    for fullPAM in returnAllPAMs():\n",
    "        dG_PAM = 0\n",
    "        dG_supercoiling= 0\n",
    "        for (target_sequence, targetPosition) in genome_dictionary[source][fullPAM]:\n",
    "            if target_sequence in dG_dictionary:\n",
    "                dG_exchange = dG_dictionary[target_sequence]\n",
    "            else:\n",
    "                dG_exchange = 0\n",
    "                dG_dictionary[target_sequence] = dG_exchange"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__New Loop with Reverse Dictionary__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Loop Comparision__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of times guides were repeated out of total: \n",
      "88915 / 3888558\n",
      "The final output should be 2.2865802696012247 % faster\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "repeated = 0\n",
    "\n",
    "for (source, targets) in genome_dictionary.items():\n",
    "    dG_dictionary = {}\n",
    "    for fullPAM in returnAllPAMs():\n",
    "        dG_PAM = 0\n",
    "        dG_supercoiling= 0\n",
    "        for (target_sequence, targetPosition) in genome_dictionary[source][fullPAM]:\n",
    "            try:\n",
    "                target_sequence in dG_dictionary\n",
    "                dG_exchange = dG_dictionary[target_sequence]\n",
    "                repeated +=1\n",
    "            except:\n",
    "                dG_exchange = 0\n",
    "                dG_dictionary[target_sequence] = dG_exchange\n",
    "            count +=1\n",
    "\n",
    "print(\"Number of times guides were repeated out of total: \")\n",
    "                \n",
    "print(repeated , \"/\" , count)\n",
    "\n",
    "print(\"The final output should be\", repeated/count*100, \"% faster\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3min 2s ± 7.88 s per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit \n",
    "for (source, targets) in genome_dictionary.items():\n",
    "    dG_dictionary = {}\n",
    "    for fullPAM in returnAllPAMs():\n",
    "        dG_PAM = 0\n",
    "        dG_supercoiling= 0\n",
    "        for (target_sequence, targetPosition) in genome_dictionary[source][fullPAM]:\n",
    "            try:\n",
    "                dG_exchange = dG_dictionary[target_sequence]\n",
    "            except:\n",
    "                dG_exchange = QuickCalc_Exchange_Energy(guide, target_sequence)\n",
    "                dG_dictionary[target_sequence] = dG_exchange"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Current dG_exchange Function and Loop__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 s ± 16.6 ms per loop (mean ± std. dev. of 10 runs, 5 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit -r 10 -n 5\n",
    "dGs = {} # I consider this equalvalent to a list  because it doesn't matter \n",
    "         # from a time percpective since it will later interate over all \n",
    "         # values anyway\n",
    "\n",
    "for (target_sequence, targetPosition) in genome_dictionary[\"Run_Genome_Plus_RC\"][\"AGGT\"]:\n",
    "    dG_exchange = QuickCalc_Exchange_Energy(guide, target_sequence)\n",
    "    dGs[target_sequence] = {'dG_exchange': dG_exchange}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Mapping the dG_exchange loop__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1e+03 ms ± 9.19 ms per loop (mean ± std. dev. of 10 runs, 5 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit -r 10 -n 5\n",
    "dG_exchanges = map(QuickCalc_Exchange_Energy, repeat(guide), (x[0] for x in genome_dictionary[\"Run_Genome_Plus_RC\"][\"AGGT\"]))\n",
    "list(dG_exchanges)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__List Comphrension on the dG_exchange loop__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.07 s ± 43.7 ms per loop (mean ± std. dev. of 10 runs, 5 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit -r 10 -n 5\n",
    "dG_exchanges = [QuickCalc_Exchange_Energy(guide, off_target[0]) for off_target in genome_dictionary[\"Run_Genome_Plus_RC\"][\"AGGT\"]]\n",
    "list(dG_exchanges)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__New dG_exchange__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {},
   "outputs": [],
   "source": [
    "# New Function\n",
    "def QuickCalc_Exchange_Energy_New(crRNA,TargetSeq):\n",
    "    dG=0\n",
    "    for i,nt in enumerate(crRNA):\n",
    "        pos=20-i\n",
    "        w1=weights[pos]\n",
    "        if nt == TargetSeq[i]:\n",
    "            continue\n",
    "        else:\n",
    "            dG+= w1*2.3\n",
    "    return float(dG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1.58135744e-05])"
      ]
     },
     "execution_count": 395,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "weights[19]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "780 ms ± 23.5 ms per loop (mean ± std. dev. of 10 runs, 5 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit -r 10 -n 5\n",
    "dGs = {}\n",
    "\n",
    "for (target_sequence, targetPosition) in genome_dictionary[\"Run_Genome_Plus_RC\"][\"AGGT\"]:\n",
    "    dG_exchange = QuickCalc_Exchange_Energy_New(guide, target_sequence)\n",
    "    dGs[target_sequence] = {'dG_exchange': dG_exchange}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Iterator Test__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "# New Iterator Function\n",
    "def QuickCalc_Exchange_Energy_Iterator(crRNA,TargetSeq):\n",
    "    dG=0\n",
    "    for i,nt in enumerate(crRNA):\n",
    "        pos = 20 - i\n",
    "        w1=weights[pos]\n",
    "        if nt == TargetSeq[i]:\n",
    "            dG+= w1*0\n",
    "        else:\n",
    "            dG+= w1*2.3\n",
    "    yield float(dG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.08 s ± 40 ms per loop (mean ± std. dev. of 10 runs, 5 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit -r 10 -n 5\n",
    "dGs = {}\n",
    "\n",
    "for (target_sequence, targetPosition) in genome_dictionary[\"Run_Genome_Plus_RC\"][\"AGGT\"]:\n",
    "    dG_exchange = next(QuickCalc_Exchange_Energy_Iterator(guide, target_sequence))\n",
    "    dGs[target_sequence] = {'dG_exchange': dG_exchange}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can see iterators would work well if we had really large datasets, but we have a small dataset repeated mutliple times."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Numpy Approach__\n",
    "\n",
    "I had one but it failed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Formatting the Weight Array__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The current weight array is formatted strangely, and so if you are able to reformat it initially, you should be able to speed up the code. The reformat takes about 5us so it is not a problem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_weights = [weight[0]*2.3 for weight in weights][0:21]\n",
    "\n",
    "def QuickCalc_Exchange_Energy_NewWeight(crRNA,TargetSeq):\n",
    "    dG=0\n",
    "    for i,nt in enumerate(crRNA):\n",
    "        pos=20-i\n",
    "        if nt == TargetSeq[i]:\n",
    "            continue\n",
    "        else:\n",
    "            dG+=new_weights[pos]\n",
    "    return float(dG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77.6 ms ± 1.96 ms per loop (mean ± std. dev. of 10 runs, 5 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit -r 10 -n 5\n",
    "\n",
    "for (target_sequence, targetPosition) in genome_dictionary[\"Run_Genome_Plus_RC\"][\"AGGT\"]:\n",
    "    dG_exchange = QuickCalc_Exchange_Energy_NewWeight(guide, target_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is an improvement of about 12x. This should allow us to get significantly faster on the code. However, so far, no modifications have been made to actual technique of calculating dG_exchange. It takes about ~50 ms for the loop to run and apparently about ~30 ms for the weights array to be called. Both of these should be upgradable using vectorization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14 s ± 228 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "\n",
    "for (source, targets) in genome_dictionary.items():\n",
    "    for fullPAM in returnAllPAMs():\n",
    "        dG_PAM = 0\n",
    "        dG_supercoiling= 0\n",
    "        for (target_sequence, targetPosition) in genome_dictionary[source][fullPAM]:\n",
    "            dG_exchange = QuickCalc_Exchange_Energy_NewWeight(guide, target_sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Numba this shit__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 664,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def QuickCalc_Exchange_Energy_PreNumba(crRNA,TargetSeq, new_weights):\n",
    "    dG=0\n",
    "    for i in range(len(crRNA)):\n",
    "        pos=20-i\n",
    "        if crRNA[i] == TargetSeq[i]:\n",
    "            continue\n",
    "        else:\n",
    "            dG+=new_weights[pos]\n",
    "    return float(dG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 665,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "115 ms ± 1.98 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "\n",
    "for (target_sequence, targetPosition) in genome_dictionary[source][fullPAM]:\n",
    "            dG_exchange = QuickCalc_Exchange_Energy_PreNumba(guide, target_sequence, new_weights)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "nt_pos={'A':0,'T':1,'C':2,'G':3}\n",
    "num_guide = np.array([nt_pos[nt] for nt in list(guide)])\n",
    "\n",
    "\n",
    "@jit()\n",
    "def QuickCalc_Exchange_Energy_Numba(crRNA,TargetSeq, np_weights):\n",
    "    \n",
    "    dG=0\n",
    "    for i in range(len(crRNA)):\n",
    "        pos=20-i\n",
    "        if  crRNA[i] == TargetSeq[i]:\n",
    "            continue\n",
    "        else:\n",
    "            dG+=np_weights[pos]\n",
    "    return float(dG)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 849,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "508 ns ± 7.31 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "dG_exchange = QuickCalc_Exchange_Energy_Numba(num_guide, num_target, np_weights)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 760,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.03 µs ± 88.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "QuickCalc_Exchange_Energy_NewWeight(guide, target_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 864,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.49 µs ± 8.16 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "\n",
    "np_weights.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 910,
   "metadata": {},
   "outputs": [],
   "source": [
    "@jit\n",
    "def Numba_test(np_weights):\n",
    "    sum = 0\n",
    "    for i in np_weights:\n",
    "        sum += i\n",
    "    return sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 912,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "223 ns ± 8.29 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "Numba_test(np_weights)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Cell magic `%%cython` not found.\n"
     ]
    }
   ],
   "source": [
    "%%cython -a\n",
    "\n",
    "def QuickCalc_Exchange_Energy_Cython(crRNA,TargetSeq, new_weights):\n",
    "    dG=0\n",
    "    for i in range(len(crRNA)):\n",
    "        pos=20-i\n",
    "        if  crRNA[i] == TargetSeq[i]:\n",
    "            continue\n",
    "        else:\n",
    "            dG+=new_weights[pos]\n",
    "    return float(dG)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 868,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.87 µs ± 66.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "np_weights = np.array(new_weights)\n",
    "\n",
    "QuickCalc_Exchange_Energy_Cython(guide, target_sequence, new_weights)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Converting the strings into numbers and getting Numba to work on a larger scale__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 884,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.96 µs ± 79.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "num_guide = np.array([ord(char) for char in list(guide)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 863,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.01 µs ± 326 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "num_guide = np.array([nt_pos[nt] for nt in list(guide)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 892,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16 ms ± 200 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "\n",
    "np.array(genome_dictionary[source][fullPAM])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 896,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.34 ms ± 183 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "\n",
    "np_off_target_array = np.array([list[0] for list in genome_dictionary[source][fullPAM]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24.2 s ± 4.87 s per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "#Convert all off target sequences into a numpy number array\n",
    "nt_pos={'A':0,'T':1,'C':2,'G':3}\n",
    "\n",
    "for (source, targets) in genome_dictionary.items():\n",
    "    np_dictionary = {}\n",
    "    for fullPAM in returnAllPAMs():\n",
    "        off_target_array = [sequence[0] for sequence in genome_dictionary[source][fullPAM]]\n",
    "        Nucleotides = []\n",
    "        for sequence in off_target_array:\n",
    "            seq = []\n",
    "            for nt in sequence:\n",
    "                seq.append(nt_pos[nt])\n",
    "            Nucleotides.append(np.array(seq))\n",
    "        np_Nucleotides = np.array(Nucleotides)\n",
    "        np_dictionary[fullPAM] = np_Nucleotides\n",
    "\n",
    "np_weights = np.array([weight[0]*2.3 for weight in weights])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.48 s ± 47.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit \n",
    "for (source, targets) in genome_dictionary.items():\n",
    "    for fullPAM in returnAllPAMs():\n",
    "        dG_PAM = 0\n",
    "        dG_supercoiling= 0\n",
    "        for off_target_sequence  in np_dictionary[fullPAM]:\n",
    "            dG_exchange = QuickCalc_Exchange_Energy_Numba(num_guide, off_target_sequence, np_weights)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "@jit(nopython = True, parallel = True)\n",
    "def QuickCalc_Exchange_Energy_Numba_Parallel(crRNA,TargetSeq, np_weights):\n",
    "    \n",
    "    dG=0\n",
    "    for i in prange(len(crRNA)):\n",
    "        pos=20-i\n",
    "        if  crRNA[i] == TargetSeq[i]:\n",
    "            continue\n",
    "        else:\n",
    "            dG+=np_weights[pos]\n",
    "    return float(dG)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": false
   },
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
